src/flag_gems/utils/pointwise_dynamic.py

```diff
@@ -132,627 +168,536 @@ def num_output_tensors(self) -> int:
-        output_tensor_index = 0
-        for i in range(op_desc.num_outputs()):
-            parameters.append(f"out{output_tensor_index}: torch.Tensor")
-            output_tensor_index += 1
+                mapping.append(non_tensor_index)
+                non_tensor_index += 1
+        return mapping
 
-    parameters.append("**kwargs")
+    def input_index(self, idx):
+        return self._input_id[idx]
 
-    return ", ".join(parameters)
+    def __str__(self) -> str:
+        return self.signature(outputs_in_arg=False)
 
 
-def ith_parameter_for_type_promotion(op_desc: OPDesc, ith: int) -> str:
-    """Generate parameter reference for i-th type promotion rule
-    Example: in0, val0, out0
-    """
-    parameters: List[str] = []
+@dataclass
+class CodeGenConfig:
+    max_tile_size: int
+    max_grid_size: Tuple[int, int, int]
+    max_num_warps_per_cta: int
 
-    input_tensor_index = 0
-    non_tensor_index = 0
-    for i in range(op_desc.num_inputs()):
-        if i not in op_desc.ith_type_promotion_args(ith):
-            if op_desc._is_tensor[i]:
-                input_tensor_index += 1
-            else:
-                non_tensor_index += 1
-            continue
-        if op_desc._is_tensor[i]:
-            parameters.append(f"in{input_tensor_index}")
-            input_tensor_index += 1
-        else:
-            parameters.append(f"val{non_tensor_index}")
-            non_tensor_index += 1
+    prefer_block_pointer: bool
+    # TODO: add 1d tile back
+    prefer_1d_tile: bool
 
-    return ", ".join(parameters)
 
+class KernelGenerator:
+    def __init__(
+        self,
+        function_schema: FunctionSchema,
+        scalar_fn: triton.JITFunction,
+        rank: int,
+        name: str,
+    ):
+        self.fx = function_schema
+        self.fn = scalar_fn
+        self.ndim = rank
+        self.name = name
+
+        self.fn_name = scalar_fn.__name__
+        self.fn_module = scalar_fn.__module__
+
+    def gen_import_function(self, code: IndentedBuffer):
+        code.writeline(f'"""Quoted source of {self.fn_name}:')
+        code.writemultiline(self.fn.src)
+        code.writeline('"""')
+        code.newline()
 
-def parameter_ref_for_wrapper(
-    op_desc: OPDesc,
-    include_outputs: bool = False,
-    include_offset: bool = False,
-    include_kwargs: bool = False,
-) -> str:
-    """Generate parameter reference for wrapper function.
-    Example: in0, val0, out0, out0_offset
-    """
-    parameters: List[str] = []
-
-    input_tensor_index = 0
-    non_tensor_index = 0
-    for i in range(op_desc.num_inputs()):
-        if op_desc._is_tensor[i]:
-            parameters.append(f"in{input_tensor_index}")
-            input_tensor_index += 1
+    def gen_decorators(self, code):
+        code.writeline("@libentry()")
+        num_non_tensor_args = self.fx.num_non_tensor_args()
+        if num_non_tensor_args > 0:
+            # we do not specialize non tensor args since they are passed into the inlined function
```

Comment:
add support for auto tunning that we can pass tuning configs and tuning prune function?

Reply:
We can have more detailed discussion on how to generate the configs and pruning methods for configs. Maybe add a few methods to generate them.